### PR TITLE
add checkpoint, metric ..

### PR DESCRIPTION
**Pull request by ggoggori**

### CODE REVIEW

File: models/train.py [hunk 1]
```diff
@@ -14,23 +15,84 @@ def make_batch(batch):
 
 def train(model, optimizer, train_dataset, device, loss_func, val_dataset=None, epochs=1):
     """Outer training loop"""
+    history = dict(train_acc=[],train_loss=[],val_acc=[],val_loss)
```
Comment (syncdoth) [inline models/train.py]: 1. val_loss 도 list 로 해주세요
2. 그리고 지금 f1 받아오고 있으니까 val_acc 이 아니라 val_f1 으로 해주세요

File: models/train.py [hunk 2]
```diff
@@ -14,23 +15,84 @@ def make_batch(batch):
...
+        
+        if val_dataset:
+            model.eval()
+            with torch.no_grad():
+                
+                for batch in val_dataset:
+                    for key in batch.keys():
+                        batch[key] = batch[key].to(device)
 
+                    loss,acc = do_train_step(model, optimizer, loss_func, batch, is_train=False)
```
Comment (syncdoth) [inline models/train.py]: variable 이름 acc 말고 f1 으로 해주세요. 그리고 변수명 사이에 space 추가해주세요

File: models/train.py [hunk 3]
```diff
@@ -14,23 +15,84 @@ def make_batch(batch):
...
+            train_loss.append(loss)
+            train_acc.append(acc)
             # TODO: Add support for metrics, checkpointing, validation, history, etc.
-            print(f"current loss: {loss}")
+        
+        ### valid
+        val_loss = []
+        val_acc = []
+        
+        if val_dataset:
```
Comment (syncdoth) [inline models/train.py]:
```
if val_dataset:
    # do stuff..
```
보다
```
if val_dataset is None:
    continue
```
가 더 좋습니다. 특히 지금처럼 depth 가 계속 들어가는 경우에는요.

File: models/train.py [hunk 4]
```diff
@@ -14,23 +15,84 @@ def make_batch(batch):
...
+                    loss,acc = do_train_step(model, optimizer, loss_func, batch, is_train=False)
+                    print(f"validation: current loss: {loss}, current f1_score: {acc}")
+                    
+                    val_loss.append(loss)
+                    val_acc.append(acc)
+                    
+                    
+        history['train_loss'].append(np.mean(train_loss))
+        history['train_acc'].append(np.mean(train_acc))
+        history['val_loss'].append(np.mean(val_loss))
```
Comment (syncdoth) [inline models/train.py]: val_dataset 이 없는 경우, `history['val_loss'].append(np.mean(val_loss))` 가 실행되지 않도록 해줘야 할 것 같습니다.

File: models/train.py [hunk 5]
```diff
@@ -14,23 +15,84 @@ def make_batch(batch):
...
+    
     for epoch in range(epochs):
+        train_loss = []
+        train_acc = []
+        
         model.train(True)
         train_bar = tqdm(train_dataset, desc=f'[train {epoch+1}/{epochs}]')
+        
         for batch in train_bar:
+
```
Comment (syncdoth) [inline models/train.py]: for loop 사이에 스페이스를 추가하지 않는게 python coding convention 입니다.

File: models/train.py [hunk 6]
```diff
@@ -14,23 +15,84 @@ def make_batch(batch):
...
     """inner training step"""
     optimizer.zero_grad()
     class_score = model(batch["images"])
     loss = loss_func(class_score, batch["targets"])
+    
+#     predicted = torch.max(class_score.data, 1)[1]
+#     batch_corr = (predicted == batch["targets"]).sum()
+#     get accuracy
+    
+    f1_socre = f1_score(batch["targets"], predicted, average='weighted')
```
Comment (syncdoth) [inline models/train.py]: weighted 말고 'macro' 혹은 'micro' 로 부탁드립니다.

File: models/train.py [hunk 7]
```diff
@@ -14,23 +15,84 @@ def make_batch(batch):
...
+    
     if is_train:
         loss.backward()
         optimizer.step()
-    return loss
+        
+    return loss, f1_score
+
+
+def save_checkpoint(epoch,model,optimizer,filename):
```
Comment (syncdoth) [inline models/train.py]: function parameter 사이에 스페이스 추가해주세요.

**Review by syncdoth — CHANGES_REQUESTED** [review]

File: models/train.py (hunk 1 above)
Comment (ggoggori) [reply models/train.py]: 넵 피드백 감사합니닷! 👍

**Review by syncdoth — APPROVED** [review]
